feat(compass-components, compass-shell): Add resize-handle to compass-components, use in compass-shell, improve accessibility COMPASS-5084

Anemy opened the pull request:
COMPASS-5084

This PR moves the resize handle we built for the sidebar to `compass-components`. It adds to variants, vertical and horizontal for different kinds of splitters, and removes the extra outline we were showing. We now use and html `input` element with the type `range` for improved accessibility. We then use this component in `compass-shell` to give a hover state for the resizer there. Synced w/ Claudia on this, we also have a resizer in `compass-aggregations`, but we were thinking it's better for that one to more minimal styling and stay how it is.

This PR also introduces [emotion](https://emotion.sh/docs/introduction) for css in js styling.

a bit outdated v


https://user-images.githubusercontent.com/1791149/134704242-a4d03fc3-db0a-4d90-a2f6-ae9cabf1df67.mp4


https://user-images.githubusercontent.com/1791149/134704248-b09ff6bf-5b70-43f1-9897-3566318ba2e0.mp4

gribnoysup commented on packages/compass-components/src/components/resize-handle.tsx:
```diff
@@ -0,0 +1,89 @@
+  onMoveLeftPressed: () => void;
+}): React.ReactElement {
+  return (
+    // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions
+    <div
+      css={[baseResizerStyles, verticalResizerStyle]}
+      onKeyDown={(e) => {
+        if (e.key === LEFT_ARROW_KEY) {
+          onMoveLeftPressed();
+        } else if (e.key === RIGHT_ARROW_KEY) {
+          onMoveRightPressed();
+        }
+      }}
+      // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
+      tabIndex={0}
```
> I was investigating using an input html element for more accessibility, but I couldn't find an easy way for the width to be dynamically sized based on the user's viewport for the range element. We can revisit down the line if it becomes higher priority I think.

Can you explain a bit more what's the issue of using range element here? I would really don't want us to start with `compass-components` by disabling the a11y rules. Maybe we can discuss it and try to find an approach that works?

If there is no way we can follow Slacks implementation and use a range input, we should consider providing this information with better set of aria attributes (I am not sure that `separator` is a good match here, it's also non-interactive, this is probably why jsx-a11y is still not happy with this code even though you added a role), here's a [slider role description](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_slider_role) for example that might fit our use-case better

gribnoysup commented on packages/compass-shell/package.json:
```diff
@@ -54,6 +54,7 @@
   "peerDependencies": {
     "@leafygreen-ui/icon": "^11.4.0",
     "@leafygreen-ui/icon-button": "^9.1.5",
+    "@mongodb-js/compass-components": "*",
```
Having peer as `*` trips npm@7 every time you do another install in the package and it changes your prod/dev version to `*` which we really don't want to happen. We also don't need to do that `*` for peers, so I'd suggest to switch it to current version

```suggestion
    "@mongodb-js/compass-components": "^0.5.0"
```

gribnoysup reviewed (commented):
Super cool to see stuff happening in the `compass-components` repo! 🔥🔥🔥

I am a bit concerned about disabling a11y rules in the library from the get-go, maybe we can spend a bit more time trying to find a better approach there?

gribnoysup commented on packages/compass-sidebar/src/components/sidebar/sidebar.jsx:
```diff
@@ -6,11 +6,11 @@ import ReactTooltip from 'react-tooltip';
 import { AutoSizer, List } from 'react-virtualized';
 import { globalAppRegistryEmit } from '@mongodb-js/mongodb-redux-common/app-registry';
 import { Resizable } from 're-resizable';
```
Something to consider for the future maybe. Would it make sense to move this library usage in compass-components also? Seems like we are repeating a lot of pretty similar code between the two otherwise 🤔

Anemy replied:
Good thinking, would be worth generalizing when/if we add more occurrences of splitters in the app 👍

Anemy commented:
@gribnoysup synced w/ Claudia a bit, going to update these to collapse when the user is dragging them past their minimum size to a certain threshold and allow pulling the resize bar to open the components when they are collapsed.

Updated the components to use `input` and have a bit better accessibility though, so that part can be taken a look at. Might make sense to wait for the other changes to come in first though, I'll ping when they're in.

Anemy commented on packages/compass-sidebar/src/components/sidebar/sidebar.jsx:
```diff
@@ -98,21 +100,21 @@ class Sidebar extends PureComponent {
+
+  toggleCollapsed(updateWidth) {
+    // When a user drags a resizer to expand/collapse we don't want to update
+    // the width as they're controlling it.
```
I don't really like passing an argument to control this, but since this component has three separate prop calls it does when it resizes I didn't totally feel comfortable updating the collapsed state in a different function. Might be a nice thing to refactor one day. Open to other suggestions on how we could do these two toggling of the collapsed state flows (one which requires updating width, one where the user is dragging and controlling the width).

gribnoysup replied:
I might be missing something but as far as I can see it there is little difference between the two cases, they all end up updating `width` to a certain value.

How I see it, you more or less have two React states here that you care about: current `width` and previous `width` (previous one you care about only when toggle button is used), everything else doesn't really matter.

Now we need to consider collapsing logic that goes something like this: when width is less than threshold, it is collapsed, otherwise it is not. As there is nothing but width used here and it's not a very complex computation, this can be derived in render method.

When you look at it like this, collapsed state and the rendered width can be derived just from the width that is stored in React state, additionally `previousWidth` can be stored when toggle is used to collapse the sidebar. This all results in a code kinda like this:

```jsx
updateWidth(width, keepPrevWidth = false) {
  this.setState(({ width: prevWidth }) => ({ width, prevWidth: keepPrevWidth ? prevWidth : null }))
}

...

render() {
  const { width, prevWidth } = this.state;
  const isCollapsed = width < sidebarMinWidthOpened;
  const renderedWidth = isCollapsed ? sidebarWidthCollapsed : boundSidebarWidth(width);

  return (
    <div id="sidebar" style={{ width: renderedWidth }}>
      ...
      <ResizeHandler ... onChange={width => this.updateWidth(width)} />
      <button id="collapse-button" onClick={() => isCollapsed ? this.updateWidth(prevWidth) : this.updateWidth(sidebarWidthCollapsed, true)}>
        ...
  )
}
```

If we want to preserve the global emit, this can be done as an effect (or in `componentDidUpdate` as this component is not a function component) deriving the collapsed state similar to the render logic code, but I can't see it being used for anything in the code except for being stored so maybe we can just drop this logic.

Anemy commented:
@gribnoysup Chatted with Claudia on this and updated accordingly. Now we allow users to resize collapse/expand these panels, also removed the triple line resizer from the shell.

Here's how collapsing/expanding with drag actions look (they feel nice :):


https://user-images.githubusercontent.com/1791149/135155194-5677cb85-efce-4d76-ad42-206e386af613.mp4


Maybe we could make the actions when it's tab selected a bit better - I think slack has it so enter opens/collapses it when it's focused, how do you feel about that? Can add it. Ran into an issue with tabbing selecting not working at all, going to investigate further, might be on main as well.

I think if we do have further uses of this component it could make sense to add a `ResizablePanel` component to `compass-components` to abstract some of the resizing/collapsing logic, but since these two are the only ones right now I think it's probably not worth it.

gribnoysup reviewed (approved):
Overall it looks superb, I really like how the interaction with this resizer feels and that they are now consistent between shell and sidebar! Left some thoughts on the `ResizeHandler` implementation, but nothing too important, just a few things that caught my eye.

Also left some thoughts on how the `isCollapsed` handling can be done, wdyt?

Not exactly related to your `toggleSidebar` method, but still kinda related and just want to highlight this separately: controlling `isCollapsed` in sidebar from `compass-home` state but keeping the `width` state in the component is a recipe for problems I think, `isCollapsed` is a derivative of width and as such syncing it back and forth especially through a separate state sounds like something that we should try to avoid. I checked `compass-home` and I don't think it does anything meaningful with this state (it switches one class name but it does nothing it seems?) so I would highly suggest to remove this state from compass-home (and not rely on `isCollapsed` being state/props in general, see my other comment about how I think we can handle this state in my opinion). There is also a global even emitted with the sidebar state, but even though the state of sidebar collapsing is preserved it doesn't look like it's meaningfully used anywhere, so I'd suggest to remove this also.

Anemy commented on packages/compass-components/src/components/resize-handle.tsx:
```diff
@@ -0,0 +1,168 @@
+  minWidth: number;
+  maxWidth: number;
+}): React.ReactElement {
+  const [isDragging, setDragging] = useState(false);
```
Ah thanks, makes total sense, definitely a little new to hooks still :)

Anemy commented on packages/compass-components/src/components/resize-handle.tsx:
```diff
@@ -0,0 +1,168 @@
+  );
+}
+
+function ResizeHandleHorizontal({
```
Updated, now they use a shared component (less duplication).

Anemy commented on packages/compass-components/src/components/resize-handle.tsx:
```diff
@@ -0,0 +1,168 @@
+});
+
+const verticalResizerStyle = css({
+  width: '4px !important',
```
Without it one of the bootstrap classes takes over, we could override that selector with one that doesn't use !important, but we'd need to do some wrapping of the input. Not much of a fan of using `!important` but I think this is a pretty scoped usage of it.

![Screen Shot 2021-09-29 at 11 57 47 AM](https://user-images.githubusercontent.com/1791149/135307309-4b97000b-f910-4fbf-9adf-1a8271e692c2.png)

gribnoysup replied:
Hmmm, got it, yeah it's very scoped, I just somehow didn't notice this issue when running locally, thanks for explaining! (As a side-note it might be something misconfigred in the build, might be worth looking into, it's a bit weird that bootstrap styles are applied after leafygreen ones)

Anemy commented on packages/compass-sidebar/src/components/sidebar/sidebar.jsx:
```diff
@@ -151,34 +153,26 @@ class Sidebar extends PureComponent {
     if (this.props.isCollapsed) {
+      if (newWidth > sidebarMinWidthOpened) {
+        // When we are not expanded and the user drags over a threshold we expand.
+        this.toggleCollapsed(false);
```
Updated the state to be more straight forward following your suggestion :)

Also moved a couple things away from being on the store which we didn't need on the store.

Anemy commented:
@gribnoysup Updated the way we're using width state in the two components (collapsed is now just derived at render) and removed some things from their store that didn't need to be there. Also cut down on a bit of the duplication in the ResizeHandle component. Thanks for the great suggestions.

gribnoysup commented on packages/compass-home/src/components/home/home.jsx:
```diff
@@ -9,9 +9,6 @@ import UI_STATES from '../../constants/ui-states';
 
```
Love this clean-up 👏

gribnoysup reviewed (approved):
Awesome work!
